File: js/heapsort.js
// Copyright © 2014 Lawrence E. Bakst. All rights reserved.
// A simple HeapSort algorithm.
var HeapSort = (function () {
    function siftup(v, ni, ri, compare) {
        var t;
        while (ri < Math.floor((ni + 1) / 2)) {
            var ci = ri * 2 + 1;                        // calculate left child node
            if (ci < ni && compare(v[ci], v[ci+1])) {   // follow the largest node left or right
                ci++;
            }
            if (compare(v[ci], v[ri]))  { // invariant holds
                break
            }
            // invariant doesn't hold, swap child and root and descent to the next level of tree
            t=v[ri], v[ri]=v[ci], v[ci]=t;
            ri = ci;
        }
    }

    return function (v, compare) {
        var n = v.length -1, t;

        //console.log("HeapSort: v.length=%o", v.length)
        // build a heap with max element at index 0 by building a binary tree and repeatedly satisfying the invaraint from the bottom up
        for (var i = Math.floor(n/2); i >= 0; i--) {
            siftup(v, n, i, compare)
        }

        while (n > 0) {
            t=v[0], v[0]=v[n], v[n]=t;
            n--;
            siftup(v, n, 0, compare);
         }
        return v;
    };
}());

/*
function main() {
    var compare = function(a, b) {
        if (a < b) {
            return true
        } else {
            return false
        }
    }

    var recs = [
        "AsfAGHM5om  00000000000000000000000000000000  0000222200002222000022220000222200002222000000001111",
        "~sHd0jDv6X  00000000000000000000000000000001  77779999444488885555CCCC777755555555BBBB666644446666",
        "uI^EYm8s=|  00000000000000000000000000000002  CCCCFFFF777799995555FFFF11112222999988884444DDDDFFFF",
        "Q)JN)R9z-L  00000000000000000000000000000003  FFFF111100000000000066668888BBBB33333333AAAA1111CCCC",
        "o4FoBkqERn  00000000000000000000000000000004  7777AAAABBBBBBBB22224444444499995555BBBB11118888DDDD",
        "*}-Wz1;TD-  00000000000000000000000000000005  AAAA88883333BBBB888888884444777722227777999900002222",
        "0fssx}~[oB  00000000000000000000000000000006  FFFF999977774444AAAA7777EEEEDDDDAAAAAAAA99998888BBBB",
        "mz4VCN@a#\"  00000000000000000000000000000007  DDDDBBBB1111FFFF2222DDDDFFFFBBBBFFFF6666444477778888",
        "my+=5r7(N|  00000000000000000000000000000008  22226666CCCC66662222FFFF0000EEEE11118888444455559999",
        "5H\\z%qt{%  00000000000000000000000000000009  0000AAAA8888FFFF0000888800000000222255551111FFFFEEEE"
    ]

    var nums = [60, 94, 66, 44, 43, 68, 7, 16, 10, 30, 52, 81, 22, 38, 32];
    console.log(HeapSort(nums, compare));
    console.log(HeapSort(recs, compare));
}

main();
*/

File: c/heapsort.c
// Copyright © 2014 Lawrence E. Bakst. All rights reserved.
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "heapsort.h"

void
siftup(Interface *d, Int ni, Int ri) {
	Int ci;

	//s := (d).(IntSlice)
	//r := s[ari]

	//    ri := ari
	//fprintf(stderr, "h(s, ni=%ld, ri=%ld), r=%ld\n", ni, ri, (Int)(d->data[ri]));
	while (ri < (ni + 1)/2) {
		ci = ri*2 + 1; // calculate left child node
		//fprintf(stderr, "ci=%ld, ri=%ld, ni=%ld\n", ci, ri, ni);

		if (ci < ni && d->Less(d, ci, ci+1)) { // follow the largest node left or right
			ci++;
			//fprintf(stderr, "ci=%ld\n", ci);
		}
		if (d->Less(d, ci, ri)) // invariant holds // ni < 1 || 
			break;
		// invariant doesn't hold, swap child and root and descent to the next level of tree
		//fprintf(stderr, "descend: Swap(%ld, %ld) = (%ld, %ld)\n", ri, ci, (Int)(d->data[ri]), (Int)(d->data[ci]));
		d->Swap(d, ri, ci);
		ri = ci;
	};
};

// Heapsort sorts data using a heaport algorithm. It implements the standard Go sort.Interface interface.
void
HeapSort(Interface *d) {
	Int i, n = d->Len(d) - 1;
	//fprintf(stderr, "HeapSort: Len=%d\n", d->Len(d));

	// build a heap with max element at index 0 by building a binary tree and repeatedly satisfying the invaraint from the bottom up
	for (i = n / 2; i >= 0; i--) {
		siftup(d, n, i);
		//Pl(d, "i");
	};
	//Pl(d, "m");
	//Pt(d)

	// repeatedly put the root (largest element) at the end of the slice and rebuild the heap with one less element
	while (n > 0) {
		//fmt.Printf("Heapsort: Swap(%d, %d), Swap(%d, %d)\n", 0, n, s[0], s[n])
		d->Swap(d, 0, n--);
		siftup(d, n, 0);
		//Pl(d, "p");
		//Pt(d)
	};
};

Int
Len(Interface *d) {
	return d->len;
};

void
Swap(Interface *d, Int i, Int j) {
	void *t;
	t = d->data[i]; d->data[i] = d->data[j]; d->data[j] = t;
};

void*
Get(Interface *d, Int i) {
	if (i < 0 || i >= d->len)
		return 0;
	return d->data[i];
}

Int
MaxLen(Interface *d) {
	return d->maxLen;
}

bool
IntLess(Interface *d, Int i, Int j) {
	if ((Int)d->data[i] < (Int)d->data[j])
		return true;
	else
		return false;
};

bool
IntAdd(Interface *d, void* p, Int len) {
	//fprintf(stderr, "d->dataLen=%ld, d->len=%ld, d->maxLen=%ld\n", d->dataLen, d->len, d->maxLen);
	d->data[d->len++] = p;
	return false;
}

Interface *
NewIntInterface(Int n) {
	Interface *d;
	if ( (d = (Interface *)calloc(sizeof(Interface), 1)) == NULL) {
		return 0;
	}
	d->Len = Len;
	d->Less = IntLess;
	d->Swap = Swap;
	d->Get = Get;
	d->Add = IntAdd;
	d->MaxLen = MaxLen;
	d->data = (void *)calloc(sizeof(void *), n);
	if (d->data == NULL) {
		return 0;
	};
	d->len = 0;
	d->dataLen = 0; // 0 says data[i] are not pointers to data but the values themselves.
	d->maxLen = n;
	return d;
}

Str
NewStr(Int len) {
	Str s;

	if ( (s = calloc(len, 1)) == NULL) {
		return 0;
	}
	return (Str)s;
}

bool
StrLess(Interface *d, Int i, Int j) {
	if (strncmp((Str)d->data[i], (Str)d->data[j], d->dataLen) < 0)
		return true;
	else
		return false;
};

bool
StrAdd(Interface *d, void* p, Int len) {
	//fprintf(stderr, "d->dataLen=%ld, d->len=%ld, d->maxLen=%ld\n", d->dataLen, d->len, d->maxLen);
	if (d->len >= d->maxLen)
		return true;
	if (len > d->dataLen)
		return true;
	d->data[d->len++] = p;
	return false;
}

Interface *
NewStrInterface(Int n, Int dataLen) {
	Interface *d;
	if ( (d = (Interface *)calloc(sizeof(Interface), 1)) == NULL) {
		return 0;
	}
	d->Len = Len;
	d->Less = StrLess;
	d->Swap = Swap;
	d->Add = StrAdd;
	d->Get = Get;
	d->data = (void *)calloc(sizeof(void *), n);
	if (d->data == NULL) {
		return 0;
	};
	d->len = 0;
	d->dataLen = dataLen;
	d->maxLen = n;
	return d;
}

void
DelInterface(Interface* d) {
	Int i;

	//fprintf(stderr, "d->Len(d)=%ld, d->MaxLen(d)=%ld, dataLen=%ld\n", d->Len(d), d->MaxLen(d), d->dataLen);

	// first if elements are pointers to data, free them
	if (d->dataLen > 0) {
		for (i = 0; i < d->Len(d); i++)
			free(d->Get(d, i));
	}

	// Now free the data array
	free(d->data);

	// Finally free the Interface
	free((void *)d);
}

